### PR TITLE
ServiceLoader loaded transformations

### PR DESCRIPTION
**Pull request by gregschohn**
### Description
Allow users to supply runtime-specified and configured transformations.  This allows users to both implement their own transformations via plugins that are loaded from ServiceLoader.  It also allows users to specify configurations of transformers that will be configured and chained together.

```
--transformer-config: Json configuration of message transformers.  Keys are the names of the loaded transformers (shortname or longname) and values are the configuration passed to each of the transformers.
```

also - here's the required syntax for the configuration for a Jolt transformation...
```
org.opensearch.migrations.transform.JsonJoltTransformerProvider expects the incoming configuration to be a Map<String,Object> or a List<Map<String,Object>>.  Each of the Maps should have one key-value, either "canned" or "script".  Canned values should be a string that specifies the name of the pre-built transformation to use addGzip,makeChunked,passThru.  Script values should be a fully-formed inlined Jolt transformation in json form.  All of the values (canned or inlined) within a configuration will be concatenated into one chained Jolt transformation
```

example scripts to configure jolt transforms, WHEN THE JOLT TRANSFORMER PROVIDER IS BEING LOADED (which it isn't by default for the ReplayTransformer as it is packaged in the docker images), would look like this...
```
[{"org.opensearch.migrations.transform.JsonJoltTransformerProvider": { "canned": "ADD_GZIP" }}]
```
or 
```
[{"org.opensearch.migrations.transform.JsonJoltTransformerProvider": { "canned": "ADD_GZIP" }},
{ "org.opensearch.migrations.transform.JsonJoltTransformerProvider":  {"script": 
    { "operation": "modify-overwrite-beta", "spec":       { "headers": {"newHeader": "newValue"}}}}}]
```

* Category: Enhancement / New feature
* Why these changes are required? Because not all migrations are created the same and different ones will have different transformation needs.
* What is the old behavior before changes and new behavior after changes? Static transformations before.  Now users can hot load and configure different (& multiple) ones.

### Issues Resolved
https://opensearch.atlassian.net/browse/MIGRATIONS-1378

### Testing
Unit testing & dockerSolution OSB w/ 4 data sets (all docs go through)

### Check List
- [x] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: TrafficCapture/trafficReplayer/README.md [hunk 1]
```diff
@@ -76,12 +76,57 @@ target URI.
...
 
-Transformations are performed via a simple class defined by
-[IJsonTransformer](../trafficReplayer/src/main/java/org/opensearch/migrations/transform/IJsonTransformer.java). Currently,
-this class uses [JOLT](https://github.com/bazaarvoice/jolt) to perform transforms that are composed of modular
-operations that are defined in the [resources](../trafficReplayer/src/main/resources/jolt/operations) associated with
-the package. Future work will include adding more JSON transformations and other potential JSON transformation tools
-(like [JMESPath](https://jmespath.org/)).
+Transformations are performed via a simple interface defined by
+[IJsonTransformer](../replayerPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/IJsonTransformer.java) ('transformer').  They are loaded dynamically and are designed to allow for easy extension
+of the TrafficReplayer to support a diverse set of needs.
```
Comment (sumobrian) [inline TrafficCapture/trafficReplayer/README.md:81]: I would indicate that this is key to supporting migration of older workloads or workloads when source and target differ with breaking API changes.

File: TrafficCapture/trafficReplayer/README.md [hunk 2]
```diff
@@ -76,12 +76,57 @@ target URI.
...
+
+The input to the transformer will be an HTTP message represented as a json-like Map<String,Object> with
+top-level key-value pairs defined in
+[JsonKeysForHttpMessage.java](../replayerPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/JsonKeysForHttpMessage.java)
```
Comment (sumobrian) [inline TrafficCapture/trafficReplayer/README.md:85]: Missing period.

File: TrafficCapture/trafficReplayer/README.md [hunk 3]
```diff
@@ -76,12 +76,57 @@ target URI.
...
+Only bodies that are json-formatted will be accessible, and they will be accessible as a fully-parsed Map (at 
+the keypath `'payload'->'inlinedJsonBody'`).  Transformers have the option to rewrite none, or any of the keys and
+values within the original message.  The transformer can return either the original message or a completely new message.
+Transformers may be used simultaneously from concurrent threads over the lifetime of the replayer.  However, 
+a message will only be processed by one transformer at a time.
+
+Transformer implementations are loaded via [Java's ServiceLoader](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html)
+by loading a jarfile that implements the [IJsonTransformerProvider]
```
Comment (sumobrian) [inline TrafficCapture/trafficReplayer/README.md:93]: Remove the new line b/w ']' and '('.

File: TrafficCapture/trafficReplayer/README.md [hunk 4]
```diff
@@ -76,12 +76,57 @@ target URI.
...
+package includes [JsonCompositeTransformer.java]
+(../replayerPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/JsonCompositeTransformer.java),
+which run transformers in serial.  That composite transformer is also utilized by the TrafficReplayer to combine the
+list of loaded transformations with a transformer to rewrite the 'Host' header.  That host transformation changes the 
+host header of every HTTP message to use the target domain-name rather than the source's.  That will be run after
+all loaded/specified transformations.
+
+Currently, there are multiple, nascent implementations included in the repository.  The 
+[jsonJoltMessageTransformerProvider](../replayerPlugins/jsonMessageTransformers/jsonJoltMessageTransformerProvider)
+package uses [JOLT](https://github.com/bazaarvoice/jolt) to perform transforms.  Some simple transformations are
```
Comment (sumobrian) [inline TrafficCapture/trafficReplayer/README.md:120]: I would reference simple transforms as examples.

**Review by sumobrian — CHANGES_REQUESTED** [review]

**Review by sumobrian — APPROVED** [review]

File: TrafficCapture/replayerPlugins/jsonMessageTransformers/jsonJoltMessageTransformerProvider/src/main/java/org/opensearch/migrations/transform/JsonJoltTransformerProvider.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                }
+            } else {
+                throw new IllegalArgumentException(getConfigUsageStr());
+            }
+            for (var c : configs) {
+                if (c.size() != 1) {
+                    throw new IllegalArgumentException(getConfigUsageStr());
+                }
+                var cannedValue = c.get(CANNED_KEY);
+                var scriptValue = c.get(SCRIPT_KEY);
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/replayerPlugins/jsonMessageTransformers/jsonJoltMessageTransformerProvider/src/main/java/org/opensearch/migrations/transform/JsonJoltTransformerProvider.java:33]: Fixed

File: TrafficCapture/replayerPlugins/jsonMessageTransformers/jsonJoltMessageTransformerProvider/src/main/java/org/opensearch/migrations/transform/JsonJoltTransformerProvider.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        return this.getClass().getName()+" expects the incoming configuration " +
+                "to be a Map<String,Object> or a List<Map<String,Object>>.  " +
+                "Each of the Maps should have one key-value, either \"canned\" or \"script\".  " +
+                "Canned values should be a string that specifies the name of the pre-built transformation to use " +
+                Arrays.stream(JsonJoltTransformBuilder.CANNED_OPERATION.values())
+                        .map(e->e.toString()).collect(Collectors.joining(",")) +
+                ".  " +
+                "Script values should be a fully-formed inlined Jolt transformation in json form.  " +
+                "All of the values (canned or inlined) within a configuration will be concatenated " +
+                "into one chained Jolt transformation";
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/replayerPlugins/jsonMessageTransformers/jsonJoltMessageTransformerProvider/src/main/java/org/opensearch/migrations/transform/JsonJoltTransformerProvider.java]: fixed

File: TrafficCapture/trafficReplayer/README.md [hunk 7]
```diff
@@ -76,12 +76,57 @@ target URI.
 
 ## Transformations
 
-Transformations are performed via a simple class defined by
-[IJsonTransformer](../trafficReplayer/src/main/java/org/opensearch/migrations/transform/IJsonTransformer.java). Currently,
-this class uses [JOLT](https://github.com/bazaarvoice/jolt) to perform transforms that are composed of modular
-operations that are defined in the [resources](../trafficReplayer/src/main/resources/jolt/operations) associated with
-the package. Future work will include adding more JSON transformations and other potential JSON transformation tools
-(like [JMESPath](https://jmespath.org/)).
+Transformations are performed via a simple interface defined by
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/README.md:79]: fixed

File: TrafficCapture/trafficReplayer/README.md (hunk 1 above)
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/README.md:81]: With the offloading of Transformers, the TrafficReplayer is now pretty much a completely workload agnostic tool.  I'd rather keep its own docs agnostic as well.

File: TrafficCapture/trafficReplayer/README.md [hunk 8]
```diff
@@ -76,12 +76,57 @@ target URI.
...
+[IJsonTransformer](../replayerPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/IJsonTransformer.java) ('transformer').  They are loaded dynamically and are designed to allow for easy extension
+of the TrafficReplayer to support a diverse set of needs.
+
+The input to the transformer will be an HTTP message represented as a json-like Map<String,Object> with
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/README.md]: done

File: TrafficCapture/trafficReplayer/README.md (hunk 2 above)
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/README.md]: done

File: TrafficCapture/trafficReplayer/README.md (hunk 3 above)
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/README.md]: done

**Pull request merged**